src/driver/hermes_cdev.c

```diff
@@ -40,8 +41,121 @@ static struct class *hermes_class;
 DEFINE_IDA(hermes_ida);
 static dev_t hermes_devt;
 
+static long handle_xdma(struct hermes_cmd_req *req,
+			struct hermes_pci_dev *hpdev)
+{
+	struct xdma_cdev *xcdev;
+	uint32_t slot_size;
+	long res;
+	loff_t pos;
+
+	pr_debug("opcode: 0x%x cid: 0x%x slot_type: 0x%x slot_id: 0x%x addr: 0x%llx len: 0x%x\n",
+			req->opcode, req->cid, req->xdma.slot_type,
+			req->xdma.slot_id, req->xdma.addr, req->xdma.len);
+
+	if (req->opcode == HERMES_WR)
+		xcdev = cdev_get_h2c(hpdev);
+	else
+		xcdev = cdev_get_c2h(hpdev);
+
+	if (!xcdev) {
+		pr_err("No DMA channel available");
+		return -HERMES_GENERIC_ERROR;
+	}
+
+	if (req->xdma.slot_type == HERMES_SLOT_PROG) {
+		if (req->xdma.slot_id >= hpdev->cfg.ehpslot) {
+			pr_err("Invalid program slot id: 0x%x\n",
+					req->xdma.slot_id);
+			return -HERMES_STATUS_INV_PROG_SLOT;
+		}
+		pos = hpdev->cfg.ehpsoff;
+		slot_size = hpdev->cfg.ehpssze;
+	} else if (req->xdma.slot_type == HERMES_SLOT_DATA) {
+		if (req->xdma.slot_id >= hpdev->cfg.ehdslot) {
+			pr_err("Invalid data slot id: 0x%x\n",
+					req->xdma.slot_id);
+			return -HERMES_STATUS_INV_DATA_SLOT;
+		}
+		pos = hpdev->cfg.ehdsoff;
+		slot_size = hpdev->cfg.ehdssze;
+	} else {
+		pr_err("Invalid slot type: %#x\n", HERMES_STATUS_INV_SLOT_TYPE);
+		return -HERMES_STATUS_INV_SLOT_TYPE;
+	}
+
+	if (req->xdma.len > slot_size) {
+		pr_err("DMA length greater than slot size: %#x > %#x\n",
+				req->xdma.len, slot_size);
+		return -HERMES_STATUS_NO_SPACE;
+	}
+
+	pos += slot_size * req->xdma.slot_id;
+
+	res = char_sgdma_read_write(xcdev, (char *) req->xdma.addr,
+			req->xdma.len, &pos, req->opcode == HERMES_WR);
+
+	if (res == -EFAULT) {
+		pr_err("Invalid DMA address: %#llx\n", req->xdma.addr);
+		return -HERMES_STATUS_INV_ADDR;
+	} else if (res < 0) {
+		return -HERMES_GENERIC_ERROR;
+	}
+
+	return res;
+}
+
+static int hermes_open(struct inode *inode, struct file *filp)
+{
+	struct hermes_dev *hermes;
+
+	hermes = container_of(inode->i_cdev, struct hermes_dev, cdev);
+	filp->private_data = hermes;
+
+	return 0;
+}
+
+static long hermes_ioctl(struct file *filp, unsigned int cmd, unsigned long arg)
+{
+	struct hermes_dev *hermes = filp->private_data;
+	struct hermes_pci_dev *hpdev = hermes->hpdev;
+	struct hermes_cmd_req_res *req_res = (struct hermes_cmd_req_res *) arg;
+	struct hermes_cmd_req *req = &req_res->req;
+	struct hermes_cmd_res *res = &req_res->res;
+	long ret = 0;
+
+	switch(req_res->req.opcode) {
+	case HERMES_WR:
+	case HERMES_RD:
+		ret = handle_xdma(req, hpdev);
+		if (ret >= 0) {
+			res->xdma.bytes = ret;
+			res->status = HERMES_STATUS_SUCCESS;
+		} else {
+			res->status = -ret;
+		}
+		ret = 0;
+		break;
+	case HERMES_REQ_SLOT:
+	case HERMES_REL_SLOT:
```

Comment:
I'd strongly avoid giving userspace access to slots. Far too many security implications and potential for abuse.